Optionally delay replica deletion

tw4l opened the pull request:
Fixes #2170

The number of days to delay file replication deletion by is configurable in the Helm chart with `replica_deletion_delay_days` (set by default to 7 days in `values.yaml` to encourage good practice, though we could change this).

When `replica_deletion_delay_days` is set to an int above 0, when a delete replica job would otherwise be started as a Kubernetes BatchJob, a cronjob is created instead with a cron schedule set to run yearly, starting x days from the current moment. This cronjob is then deleted by the operator after the job successfully completes. If a failed background job is retried, it is re-run immediately as a BatchJob rather than being scheduled out into the future again.

This requires manual testing due to the amount of time involved, which would best be done on dev.

tw4l committed:
WIP: Implement delayed deletion of replica files

tw4l committed:
Delete replica deletion cron job after one success

tw4l committed:
Linting and typing fixups

tw4l committed:
Document replica_deletion_delay_days

ikreymer commented on backend/btrixcloud/crawlmanager.py:
```diff
@@ -393,3 +396,37 @@ async def update_scheduled_job(
+        """create scheduled job to delay replica file in x days"""
+        now = dt_now()
+        run_at = now + timedelta(days=delay_days)
+        schedule = f"0 0 {run_at.day} {run_at.month} *"
```
```suggestion
        schedule = f"{run_at.minute} {run_at.hour} {run_at.day} {run_at.month} *"
```
I think this should be a full day from moment of deletion, to avoid scenarios where its still deleted quickly if its at the end of the day.

ikreymer committed:
Update backend/btrixcloud/crawlmanager.py

ikreymer committed:
cronjob needs to specify labels for spawned job as well
add 'job-id' label instead of 'name', which is automatically set, use that if available

ikreymer committed:
keep 'replica_deletion_delay_days' to 0 by default

ikreymer commented:
Looks good! Made some small tweaks: added a fix to ensure Job gets the correct labels from CronJob, and set the time to be in exact 24-hour durations from deletion time, and set default to 0 (for now), so those deploying on their own don't end up with unexpected data..
